### PR TITLE
CDAP-7109 Specify that ds access can be done from Partitioner/Comparator classes.

### PR DESCRIPTION
**Pull request by anwar6953**
CDAP-7109 Specify that ds access can be done from Partitioner/Comparator classes.

https://issues.cask.co/browse/CDAP-7109
Also resolves https://issues.cask.co/browse/CDAP-3544, because it had been mostly resolved before anyways.

Docs quick build: http://builds.cask.co/browse/CDAP-DQB124-1
MapReduce Programs page: http://builds.cask.co/artifact/CDAP-DQB124/shared/build-1/Docs-HTML/3.5.1-SNAPSHOT/en/developers-manual/building-blocks/mapreduce-programs.html

### CODE REVIEW

File: cdap-docs/developers-manual/source/building-blocks/mapreduce-programs.rst [hunk 1]
```diff
@@ -136,7 +136,8 @@ MapReduce and Datasets
 .. rubric: Reading and Writing to Datasets from a MapReduce program
 
 Both a CDAP ``mapper`` and ``reducer`` can directly read or write to a dataset, using
-one of these options:
+one of the following options. Note that options 2 and 3 can also be used for a Partitioner
+and Comparator class, if configured on the MapReduce job:
```
Comment (anew) [inline cdap-docs/developers-manual/source/building-blocks/mapreduce-programs.rst]: to be consistent with the first sentence, should it be `partitioner` and `comparator` (without class)?

**Comment by anew** [comment]
should we have an example for this?

**Comment by anwar6953** [comment]
I didn't add an example, because its the same as Mapper/Reducer (the initialize/destroy methods), with the only difference being the class encapsulating the initialize/destroy methods. Because of that, I didn't feel an example was necessary.

Edit: Also, it is more of an advanced feature, so I think the utility of the example would not be worth the real estate it takes up in the documentation.

File: cdap-docs/developers-manual/source/building-blocks/mapreduce-programs.rst (hunk 1 above)
Comment (anwar6953) [reply cdap-docs/developers-manual/source/building-blocks/mapreduce-programs.rst]: I made the mapper and reducer words capitalized.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 2]
```diff
@@ -298,7 +298,7 @@ To put a value into a token, first obtain access to the token from the workflow
 and then set a value for a specific key. 
 
 In the case of a MapReduce program, the program's Mapper and Reducer classes need to
-implement ``ProgramLifecycle<MapReduceContext>``. After doing so, they can access the
+implement ``ProgramLifecycle<MapReduceTaskContext>``. After doing so, they can access the
```
Comment (awholegunch) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst:301]: Is this an API change? I don't think that this has been documented adequately.

Comment (anwar6953) [reply cdap-docs/developers-manual/source/building-blocks/workflows.rst:301]: It's an API change that happened previously (around 3.4.0). It was backwards-compatible, though.
What's missing, do you think?

File: cdap-docs/developers-manual/source/building-blocks/mapreduce-programs.rst [hunk 3]
```diff
@@ -135,8 +135,9 @@ MapReduce and Datasets
 
 .. rubric: Reading and Writing to Datasets from a MapReduce program
 
-Both a CDAP ``mapper`` and ``reducer`` can directly read or write to a dataset, using
-one of these options:
+Both a CDAP ``Mapper`` and ``Reducer`` can directly read or write to a dataset, using
+one of the following options. (Note that the second and third options can be used for a
+`Partitioner` or `Comparator` class, if configured on the MapReduce job.)
```
Comment (awholegunch) [inline cdap-docs/developers-manual/source/building-blocks/mapreduce-programs.rst]: "Slack" mistake:

```
``Partitioner`` or ``Comparator`` class, if configured on the MapReduce job.)
```

Comment (anwar6953) [reply cdap-docs/developers-manual/source/building-blocks/mapreduce-programs.rst]: fixed

**Comment by awholegunch** [comment]
LGTM

File: cdap-docs/developers-manual/source/building-blocks/mapreduce-programs.rst [hunk 4]
```diff
@@ -135,9 +135,9 @@ MapReduce and Datasets
 
 .. rubric: Reading and Writing to Datasets from a MapReduce program
 
-Both a CDAP ``mapper`` and ``reducer`` can directly read or write to a dataset, using
-one of the following options. Note that options 2 and 3 can also be used for a Partitioner
-and Comparator class, if configured on the MapReduce job:
+Both a CDAP ``Mapper`` and ``Reducer`` can directly read or write to a dataset, using
+one of the following options. (Note that the second and third options can be used for a
+``Partitioner`` or ``Comparator`` class, if configured on the MapReduce job.)
```
Comment (anew) [inline cdap-docs/developers-manual/source/building-blocks/mapreduce-programs.rst]: still should remove the word "class"

**Comment by anew** [comment]
LGTM

**Comment by anwar6953** [comment]
Squashed commits; merging.

**Pull request merged**